[CT-2636] TestVersionedModels::test_pp_versioned_models is flaky

Example failing test run: https://github.com/dbt-labs/dbt-core/actions/runs/5178505585/jobs/9330058721?pr=7768

Relevant Logs: 

```sh
        # update versions schema.yml block - latest_version from 2 to 1
        write_file(
            models_versions_updated_schema_yml, project.project_root, "models", "schema.yml"
        )
>       results, log_output = run_dbt_and_capture(
            ["--partial-parse", "--log-format", "json", "run"]
        )

D:\a\dbt-core\dbt-core\tests\functional\partial_parsing\test_partial_parsing.py:341: 
```

```sh
        if expect_pass is not None:
>           assert res.success == expect_pass, "dbt exit state did not match expected"
E           AssertionError: dbt exit state did not match expected

d:\a\dbt-core\dbt-core\core\dbt\tests\util.py:105: AssertionError
```

```sh
stdout_log: {"data": {"msg": "  relation \"test16859771223328142563_test_partial_parsing.model_one_downstream\" does not exist"}, "info": {"category": "", "code": "Z029", "extra": {}, "invocation_id": "9e30ed11-b91f-4609-8dd0-a82ab93b96ba", "level": "error", "msg": "  relation \"test16859771223328142563_test_partial_parsing.model_one_downstream\" does not exist", "name": "AfterFirstRunResultError", "pid": 6128, "thread": "MainThread", "ts": "2023-06-05T14:58:50.611377Z"}}
```

--- a/core/dbt/include/global_project/macros/materializations/models/table/table.sql
+++ b/core/dbt/include/global_project/macros/materializations/models/table/table.sql
@@ -33,7 +33,12 @@
 
   -- cleanup
   {% if existing_relation is not none %}
-      {{ adapter.rename_relation(existing_relation, backup_relation) }}
+     /* Do the equivalent of rename_if_exists. 'existing_relation' could have been dropped
+        since the variable was first set. */
+    {% set existing_relation = load_cached_relation(existing_relation) %}
+    {% if existing_relation is not none %}
+        {{ adapter.rename_relation(existing_relation, backup_relation) }}
+    {% endif %}
   {% endif %}
 
   {{ adapter.rename_relation(intermediate_relation, target_relation) }}

--- a/core/dbt/include/global_project/macros/materializations/models/view/view.sql
+++ b/core/dbt/include/global_project/macros/materializations/models/view/view.sql
@@ -45,7 +45,12 @@
   -- cleanup
   -- move the existing view out of the way
   {% if existing_relation is not none %}
-    {{ adapter.rename_relation(existing_relation, backup_relation) }}
+     /* Do the equivalent of rename_if_exists. 'existing_relation' could have been dropped
+        since the variable was first set. */
+    {% set existing_relation = load_cached_relation(existing_relation) %}
+    {% if existing_relation is not none %}
+        {{ adapter.rename_relation(existing_relation, backup_relation) }}
+    {% endif %}
   {% endif %}
   {{ adapter.rename_relation(intermediate_relation, target_relation) }}
 

